asf_heat_pump_suitability/notebooks/reweight_epc_test.py

```diff
@@ -0,0 +1,272 @@
+        "Flat, maisonette or apartment",
+        "A caravan or other mobile or temporary structure",
+    ]
+)
+
+int_cols = ["owner-occupied", "rental (private)", "rental (social)"]
+tenure = tenure.with_columns(
+    [pl.col(int_cols).str.replace("c", f"{censored_vals}").cast(pl.Int64)]
+).select(pl.col(["lsoa", "owner-occupied", "rental (private)", "rental (social)"]))
+
+# %% [markdown]
+# ## Get marginals
+
+# %%
+# Choose test LSOA
+test_lsoa = "E01002058"  # small sample
+# test_lsoa = "W01000527" # good example
+# test_lsoa = "E01012376"  # missing val
+# test_lsoa = "W01000328"  # example in slides
+# test_lsoa = "E01033942"  # example on slack
+
+# %%
+# Proportions of tenure in census data
+tenure_subset = tenure.filter(pl.col("lsoa") == test_lsoa)
+tenure_dict = tenure_subset.to_dicts()[0]
+del tenure_dict["lsoa"]
+
+# %%
+# Proportions of property_type in census data
+property_type_subset = property_type.filter(pl.col("lsoa") == test_lsoa)
+property_dict = property_type_subset.to_dicts()[0]
+del property_dict["lsoa"]
+
+# %%
+# Generate artificial target population dataset with proportions derived from census data
+marginals = {
+    "tenure": {k: v / sum(tenure_dict.values()) for k, v in tenure_dict.items()},
+    "property_type": {
+        k: v / sum(property_dict.values()) for k, v in property_dict.items()
+    },
+}
+target_df = balance.weighting_methods.rake.prepare_marginal_dist_for_raking(marginals)
+target = balance.Sample.from_frame(target_df)
+
+# %% [markdown]
+# ## Add dummy rows to sample subset
+
+# %%
+subset = sample.filter(pl.col("lsoa") == test_lsoa)
+
+# %%
+target_features = {"tenure": tenure_dict, "property_type": property_dict}
+
+
+# %%
+def generate_df_dummies(feature_dicts):
+    # Identify categories present in target but not sample
+    dummies = []
+    for feature, _dict in feature_dicts.items():
+        _relevant_cats = {
+            k for k in _dict.keys() if _dict.get(k) > 0
+        }  # only add dummy rows for missing categories where target has observations but sample doesn't
+        _missing_cats = _relevant_cats.difference(set(subset[feature].unique()))
+        dummies.append({feature: list(_missing_cats)})
+    return [pl.DataFrame(d) for d in dummies]
+
+
+# %%
+_dfs = generate_df_dummies(target_features)
+
+dummy_rows = pl.concat(_dfs, how="horizontal").with_columns(
+    pl.lit(test_lsoa, pl.String).alias("lsoa")
+)
+dummy_rows = dummy_rows.with_columns(
+    pl.Series(name="id", values=[f"dummy_{_}" for _ in range(0, len(dummy_rows))])
+)  # create dummy df
+
+
+# Fill in null values in dummy df
+for feature in ["tenure", "property_type"]:
+    missing_vals = list(dummy_rows.drop_nulls()[feature].unique())
+    if dummy_rows[feature].is_null().sum() > 0:
+        if len(missing_vals) > 0:  # fill with random missing category first
+            dummy_rows = dummy_rows.with_columns(
+                pl.col(feature).fill_null(random.choices(missing_vals, k=1)[0])
+            )
```

Comment:
I guess this is going to fill in all nulls with the same one random missing category, would be nicer to be all random?

This doesn't quite work, but something like:
```
dummy_rows.with_columns(
    pl.when(pl.col(feature).is_null())
    .then(pl.lit(random.choices(missing_vals,k=len(dummy_rows))))
    .otherwise(pl.col(feature))
    .alias('Prob')
)
```

Reply:
this will have to be addressed in a separate issue as i still can't get it to work at the moment